### PR TITLE
Bug 1280460 - Fix Today Widget Layout issues

### PR DESCRIPTION
**Pull request by farhanpatel**
There are some layout issues that appear with the Today widget. Understanding the layout here was getting very tricky. Now that we are ios9+, I ended up rearranging the views using UIStackview which in my opinion simplifies things. I still want to do some more testing. Just opening the PR for now.

### CODE REVIEW

File: Extensions/Today/TodayViewController.swift [hunk 1]
```diff
@@ -9,21 +9,20 @@ import SnapKit
 
 private let log = Logger.browserLogger
 
+struct TodayStrings {
+    static let NewPrivateTabButtonLabel = NSLocalizedString("TodayWidget.NewPrivateTabButtonLabel", value: "New Private Tab", tableName: "Today", comment: "New Private Tab button label")
+    static let NewTabButtonLabel = NSLocalizedString("TodayWidget.NewTabButtonLabel", value: "New Tab", tableName: "Today", comment: "New Tab button label")
+    static let GoToCopiedLinkLabel = NSLocalizedString("TodayWidget.GoToCopiedLinkLabel", value: "Go to copied link", tableName: "Today", comment: "Go to link on clipboard")}
```
Comment (thebnich) [inline Extensions/Today/TodayViewController.swift]: Nit: newline for trailing `}`

File: Extensions/Today/TodayViewController.swift [hunk 2]
```diff
@@ -9,21 +9,20 @@ import SnapKit
...
     static let imageButtonTextSize: CGFloat = 14.0
-
-    static let copyLinkButtonHeight: CGFloat = 44
-
-    static let verticalWidgetMargin: CGFloat = 10
-    static let horizontalWidgetMargin: CGFloat = 10
-    static var defaultWidgetTextMargin: CGFloat = 22
-
-    static let buttonSpacerMultipleOfScreen = 0.4
+    static let coplyLinkImageWidth: CGFloat = 23
```
Comment (thebnich) [inline Extensions/Today/TodayViewController.swift]: `coply` -> `copy`

File: Extensions/Today/TodayViewController.swift [hunk 3]
```diff
@@ -80,8 +80,29 @@ class TodayViewController: UIViewController, NCWidgetProviding {
...
+        stackView.axis = .Vertical
+        stackView.alignment = .Fill
+        stackView.spacing = 0
+        stackView.distribution = UIStackViewDistribution.Fill
+        stackView.layoutMargins = UIEdgeInsets(top: TodayUX.margin, left: TodayUX.margin, bottom: TodayUX.margin, right: TodayUX.margin)
+        stackView.layoutMarginsRelativeArrangement = true
+        return stackView
+    }()
+    
+    private lazy var buttonStackView: UIStackView = {
```
Comment (thebnich) [inline Extensions/Today/TodayViewController.swift]: Nit: This is `lazy`, but `widgetStackView` isn't. For the sake of consistency, I'd make them both `lazy` or non-`lazy` (don't think it matters much either way).

File: Extensions/Today/TodayViewController.swift [hunk 4]
```diff
@@ -80,8 +80,29 @@ class TodayViewController: UIViewController, NCWidgetProviding {
...
+        let stackView = UIStackView()
+        stackView.axis = .Horizontal
+        stackView.alignment = .Fill
+        stackView.spacing = 0
+        stackView.distribution = UIStackViewDistribution.FillEqually
+        let edge = self.view.frame.size.width * TodayUX.buttonsHorizontalMarginPercentage
+        print(edge)
```
Comment (thebnich) [inline Extensions/Today/TodayViewController.swift]: Remove

File: Extensions/Today/TodayViewController.swift [hunk 5]
```diff
@@ -107,84 +128,36 @@ class TodayViewController: UIViewController, NCWidgetProviding {
...
-        newPrivateTabButton.snp_makeConstraints { make in
-            make.centerY.equalTo(newTabButton.snp_centerY)
-            make.centerX.equalTo(buttonSpacer.snp_right)
-        }
+        view.addSubview(widgetStackView)
 
-        newTabButton.label.snp_makeConstraints { make in
-            make.leading.greaterThanOrEqualTo(view)
+        widgetStackView.snp_makeConstraints { make in
+            make.top.left.right.bottom.equalTo(self.view)
```
Comment (thebnich) [inline Extensions/Today/TodayViewController.swift]: You can reduce this to `make.edges`

File: Extensions/Today/TodayViewController.swift [hunk 6]
```diff
@@ -261,19 +223,16 @@ class ImageButtonWithLabel: UIView {
...
         }
 
-        snp_makeConstraints { make in
-            make.width.equalTo(button)
-            make.height.equalTo(button)
+        label.snp_makeConstraints { make in
+            make.width.equalTo(self)
+            make.top.equalTo(button.snp_bottom)
+            make.bottom.equalTo(self.snp_bottom)
+            make.centerX.equalTo(self)
```
Comment (thebnich) [inline Extensions/Today/TodayViewController.swift]: I'd reduce this to:
```
make.top.equalTo(button.snp_bottom)
make.leading.trailing.bottom.equalTo(self)
```

File: Extensions/Today/TodayViewController.swift [hunk 7]
```diff
@@ -300,41 +258,35 @@ class ButtonWithSublabel: UIButton {
...
         let imageView = self.imageView!
-
         let subtitleLabel = self.subtitleLabel
         subtitleLabel.textColor = UIColor.lightGrayColor()
         self.addSubview(subtitleLabel)
 
         imageView.snp_remakeConstraints { make in
             make.centerY.equalTo(self.snp_centerY)
-            make.right.equalTo(titleLabel.snp_left).offset(-TodayUX.horizontalWidgetMargin)
+            make.left.equalTo(self.snp_left)
```
Comment (thebnich) [inline Extensions/Today/TodayViewController.swift]: Prefer `leading`/`trailing` over `left`/`right` for RTL compatibility.

Also, you can chain these together if you want (`make.leading.centerY`).

(Same comments for other constraint definitions below.)

File: Extensions/Today/TodayViewController.swift [hunk 8]
```diff
@@ -300,41 +258,35 @@ class ButtonWithSublabel: UIButton {
...
         self.titleLabel?.removeFromSuperview()
         addSubview(titleLabel)
 
         let imageView = self.imageView!
-
         let subtitleLabel = self.subtitleLabel
         subtitleLabel.textColor = UIColor.lightGrayColor()
         self.addSubview(subtitleLabel)
 
         imageView.snp_remakeConstraints { make in
```
Comment (thebnich) [inline Extensions/Today/TodayViewController.swift]: `remake` -> `make`

File: Extensions/Today/TodayViewController.swift [hunk 9]
```diff
@@ -300,41 +258,35 @@ class ButtonWithSublabel: UIButton {
...
             make.centerY.equalTo(self.snp_centerY)
-            make.right.equalTo(titleLabel.snp_left).offset(-TodayUX.horizontalWidgetMargin)
+            make.left.equalTo(self.snp_left)
+            make.width.equalTo(TodayUX.coplyLinkImageWidth)
+        }
+
+        titleLabel.snp_remakeConstraints { make in
```
Comment (thebnich) [inline Extensions/Today/TodayViewController.swift]: `remake` -> `make`

**Comment by farhanpatel** [comment]
@thebnich Should be good for one more look over. I think you need to approve the changes again according to Github.

**Pull request merged**